btcec/nfield.go

```diff
@@ -0,0 +1,1769 @@
+// the intermediate results while adding or multiplying two 64-bit numbers, and
+// second there is no space left for overflows when performing the intermediate
+// arithmetic between each array element which would lead to expensive carry
+// propagation.
+//
+// Given the above, this implementation represents the the field elements as
+// 10 uint32s with each word (array entry) treated as base 2^26.  This was
+// chosen for the following reasons:
+// 1) Most systems at the current time are 64-bit (or at least have 64-bit
+//    registers available for specialized purposes such as MMX) so the
+//    intermediate results can typically be done using a native register (and
+//    using uint64s to avoid the need for additional half-word arithmetic)
+// 2) In order to allow addition of the internal words without having to
+//    propagate the the carry, the max normalized value for each register must
+//    be less than the number of bits available in the register
+// 3) Since we're dealing with 32-bit values, 64-bits of overflow is a
+//    reasonable choice for #2
+// 4) Given the need for 256-bits of precision and the properties stated in #1,
+//    #2, and #3, the representation which best accomodates this is 10 uint32s
+//    with base 2^26 (26 bits * 10 = 260 bits, so the final word only needs 22
+//    bits) which leaves the desired 64 bits (32 * 10 = 320, 320 - 256 = 64) for
+//    overflow
+//
+// Since it is so important that the field arithmetic is extremely fast for
+// high performance crypto, this package does not perform any validation where
+// it ordinarily would.  For example, some functions only give the correct
+// result is the field is normalized and there is no checking to ensure it is.
+// While I typically prefer to ensure all state and input is valid for most
+// packages, this code is really only used internally and every extra check
+// counts.
+
+import (
+	"encoding/hex"
+	"fmt"
+)
+
+// Constants related to the nfield representation.
+const (
+	// nfieldWords is the number of words used to internally represent the
+	// 256-bit value.
+	nfieldWords = 10
+
+	// nfieldBase is the exponent used to form the numeric base of each word.
+	// 2^(nfieldBase*i) where i is the word position.
+	nfieldBase = 26
+
+	// nfieldOverflowBits is the minimum number of "overflow" bits for each
+	// word in the nfield value.
+	nfieldOverflowBits = 32 - nfieldBase
+
+	// nfieldBaseMask is the mask for the bits in each word needed to
+	// represent the numeric base of each word (except the most significant
+	// word).
+	nfieldBaseMask = (1 << nfieldBase) - 1
+
+	// nfieldMSBBits is the number of bits in the most significant word used
+	// to represent the value.
+	nfieldMSBBits = 256 - (nfieldBase * (nfieldWords - 1))
+
+	// nfieldMSBMask is the mask for the bits in the most significant word
+	// needed to represent the value.
+	nfieldMSBMask = (1 << nfieldMSBBits) - 1
+
+	// nfieldPrimeWordZero is word zero of the secp256k1 prime in the
```

Comment:
There are several more cases of this, but I'll not comment on every one to avoid sending a ton of e-mails.